crates/net/eth-wire/src/muxdemux.rs

```diff
@@ -0,0 +1,384 @@
+    /// open.
+    pub fn shared_capabilities(&self) -> &SharedCapabilities {
+        &self.shared_capabilities
+    }
+
+    fn shared_cap<'a>(
+        cap: &Capability,
+        shared_capabilities: &'a SharedCapabilities,
+    ) -> Result<&'a SharedCapability, MuxDemuxError> {
+        for shared_cap in shared_capabilities.iter_caps() {
+            match shared_cap {
+                SharedCapability::Eth { .. } if cap.is_eth() => return Ok(shared_cap),
+                _ => continue,
+                // more caps to come ..
+            }
+        }
+
+        Err(CapabilityNotShared)
+    }
+
+    fn reg_new_ingress_buffer(
+        &mut self,
+        cap: &SharedCapability,
+    ) -> Result<mpsc::UnboundedReceiver<BytesMut>, MuxDemuxError> {
+        if let Some(tx) = self.demux.get(cap) {
+            if !tx.is_closed() {
+                return Err(StreamAlreadyExists)
+            }
+        }
+        let (ingress_tx, ingress) = mpsc::unbounded_channel();
+        self.demux.insert(cap.clone(), ingress_tx);
+
+        Ok(ingress)
+    }
+
+    fn unmask_msg_id(&self, id: &mut u8) -> Result<&SharedCapability, MuxDemuxError> {
+        for cap in self.shared_capabilities.iter_caps() {
+            let offset = cap.relative_message_id_offset();
+            let next_offset = offset + cap.num_messages()?;
+            if *id < next_offset {
+                *id -= offset;
+
+                return Ok(cap)
+            }
+        }
+
+        Err(MessageIdOutOfRange(*id))
+    }
+
+    /// Masks message id with offset relative to the message id suffix reserved for capability
+    /// message ids. The p2p stream further masks the message id (todo: mask whole message id at
+    /// once to avoid copying message to mutate id byte or sink BytesMut).
+    fn mask_msg_id(&self, msg: Bytes) -> Bytes {
+        let mut masked_bytes = BytesMut::zeroed(msg.len());
+        masked_bytes[0] = msg[0] + self.owner.relative_message_id_offset();
+        masked_bytes[1..].copy_from_slice(&msg[1..]);
+
+        masked_bytes.freeze()
+    }
+
+    /// Checks if all clones of this shared stream have been dropped, if true then returns //
+    /// function to drop the stream.
+    fn can_drop(&mut self) -> bool {
+        for tx in self.demux.values() {
+            if !tx.is_closed() {
+                return false
+            }
+        }
+
+        true
+    }
+}
+
+impl<S, E> Stream for MuxDemuxStream<S>
+where
+    S: Stream<Item = Result<BytesMut, E>> + CanDisconnect<Bytes> + Unpin,
+    MuxDemuxError: From<E> + From<<S as Sink<Bytes>>::Error>,
+{
+    type Item = Result<BytesMut, MuxDemuxError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // sink buffered bytes from `StreamClone`s
+        for _ in 0..self.demux.len() {
+            if self.inner.poll_ready_unpin(cx).is_pending() {
+                break
+            }
```

Comment:
we should flip this:

as long as the underlying sink can accept messages, start sending buffered messages from `demux`.

Reply:
so the main stream should not be prioritised over the secondary streams? e.g. eth protocol should not be prioritised over snap protocol?

Reply:
commit [c6c86d1](https://github.com/paradigmxyz/reth/pull/5577/commits/c6c86d1c861fc00cd5c956120463fdd28d328bf6) addresses this by sending at least as many messages as there are stream clones, then sending as long as their is no message received for the main stream